crates/bevy_render/src/world_sync.rs

```diff
@@ -0,0 +1,263 @@
+/// This is called "Pipelined Rendering", see [`PipelinedRenderingPlugin`] for more information.
+///
+/// [`WorldSyncPlugin`] is the first thing that runs every frame and it maintains an entity-to-entity mapping
+/// between the main world and the render world, by spawning new entities (without any components) in the main world,
+/// and despawning entities in the render world to match newly despawned entities in the main world.
+/// This is necessary preparation for extraction ([`ExtractSchedule`](crate::ExtractSchedule)), which copies over component data from the main
+/// to the render world for these entities.
+///
+/// ```text
+/// |--------------------------------------------------------------------|
+/// |      |         |          Main world update                        |
+/// | sync | extract |---------------------------------------------------|
+/// |      |         |         Render world update                       |
+/// |--------------------------------------------------------------------|
+/// ```
+///
+/// An example for synchronized main entities 1v1 and 18v1
+///
+/// ```text
+/// |---------------------------Main World------------------------------|
+/// |  Entity  |                    Component                           |
+/// |-------------------------------------------------------------------|
+/// | ID: 1v1  | PointLight | RenderEntity(ID: 3V1) | SyncToRenderWorld |
+/// | ID: 18v1 | PointLight | RenderEntity(ID: 5V1) | SyncToRenderWorld |
+/// |-------------------------------------------------------------------|
+///
+/// |----------Render World-----------|
+/// |  Entity  |       Component      |
+/// |---------------------------------|
+/// | ID: 3v1  | MainEntity(ID: 1V1)  |
+/// | ID: 5v1  | MainEntity(ID: 18V1) |
+/// |---------------------------------|
+///
+/// ```
+///
+/// Note that this effectively establishes a link between the main world entity and the render world entity.
+/// Not every entity needs to be synchronized, however; only entities with the [`SyncToRenderWorld`] component are synced.
+/// Adding [`SyncToRenderWorld`] to a main world component will establish such a link.
+/// Once a synchronized main entity is despawned, its corresponding render entity will be automatically
+/// despawned in the next `sync`.
+///
+/// The sync step does not copy any of component data between worlds, since its often not necessary to transfer over all
+/// the components of a main world entity.
+/// The render world probably cares about a `Position` component, but not a `Velocity` component.
+/// The extraction happens in its own step, independently from, and after synchronization.
+///
+/// Moreover, [`WorldSyncPlugin`] only synchronizes *entities*. [`RenderAsset`](crate::render_asset::RenderAsset)s like meshes and textures are handled
+/// differently.
+///
+/// [`PipelinedRenderingPlugin`]: crate::pipelined_rendering::PipelinedRenderingPlugin
+#[derive(Default)]
+pub struct WorldSyncPlugin;
```

Comment:
IMO we should cover `RenderEntity` and `MainEntity` here as well, since those are important parts of the puzzle.